src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/FileRenameTracker.cs

```diff
@@ -0,0 +1,100 @@
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Linq;
+using Task = System.Threading.Tasks.Task;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IProjectChangeHintReceiver)), Export]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityRenameHint.RenamedFileAsString)]
+    [AppliesTo(ProjectCapability.CSharp)]
+    internal class FileRenameTracker : IProjectChangeHintReceiver
+    {
+        private readonly IUnconfiguredProjectVsServices _unconfiguredProjectVsServices;
+        private IComponentModel _componentModel;
+        private VisualStudioWorkspace _visualStudioWorkspace;
+
+        /// <summary>
+        /// The thread handling service.
+        /// </summary>
+        [Import]
+        private IProjectThreadingService ThreadingService { get; set; }
+
+        /// <summary>
+        /// Gets the VS global service provider.
+        /// </summary>
+        [Import]
+        protected SVsServiceProvider ServiceProvider { get; private set; }
+
+        [ImportingConstructor]
+        public FileRenameTracker(IUnconfiguredProjectVsServices projectVsServices)
+        {
+            Requires.NotNull(projectVsServices, nameof(projectVsServices));
+            _unconfiguredProjectVsServices = projectVsServices;
+        }
+
+        public async Task HintedAsync(IImmutableDictionary<Guid, IImmutableSet<IProjectChangeHint>> hints)
+        {
+            var files = hints.GetValueOrDefault(ProjectChangeFileSystemEntityRenameHint.RenamedFile) ?? ImmutableHashSet.Create<IProjectChangeHint>();
+            if (files.Count == 1)
+            {
+                var hint = files.First() as IProjectChangeFileRenameHint;
+                if (hint != null && !hint.ChangeAlreadyOccurred)
+                {
+                    var kvp = hint.RenamedFiles.First();
+                    await ScheduleRenameAsync(kvp.Key, kvp.Value).ConfigureAwait(false);
+                }
+            }
+        }
+
+        public Task HintingAsync(IProjectChangeHint hint)
+        {
+            return TplExtensions.CompletedTask;
+        }
+
+        private async Task ScheduleRenameAsync(string oldFilePath, string newFilePath)
+        {
+            string codeExtension = Path.GetExtension(newFilePath);
+            if (codeExtension == null || !oldFilePath.EndsWith(codeExtension, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            await ThreadingService.SwitchToUIThread();
+
+            if (_visualStudioWorkspace == null)
+            {
+                _componentModel = (IComponentModel)ServiceProvider.GetService(typeof(SComponentModel));
+                _visualStudioWorkspace = _componentModel.GetService<VisualStudioWorkspace>();
```

Comment:
Ah, VisualStudioWorkspace is a MEF component, yes we should import it instead of having this IServiceProvider dance.